service/src/java/org/apache/hive/service/cli/thrift/ThriftCLIService.java

```diff
@@ -147,15 +151,81 @@
 
   protected ThreadLocal<ServerContext> currentServerContext;
 
+  /**
+   * A context container for information related to client connections to the
+   * Thrift server.
+   */
   static class ThriftCLIServerContext implements ServerContext {
-    private SessionHandle sessionHandle = null;
 
-    public void setSessionHandle(SessionHandle sessionHandle) {
-      this.sessionHandle = sessionHandle;
+    private final long createTime;
+    private final Collection<SessionHandle> hiveSessionHandles;
```

Comment:
Why is this?
Would not this collect SessionHandle-s forever? Where do we clean this?

Reply:
Why would we need to keep previous sessionHandles?

Reply:
It is my understanding, that in the context of something like a JDBC Database connection pool, a connection to Hive can be re-used, in which case, a single connection can have many Session handles associated with it (sequentially, over time).  I can think of a scenario where it might be helpful to have a history size of one: if something about one session breaks something that then cascades to the next session.  I picked four arbitrarily, but this number can be adjusted once it's used a bit and see if it's helpful or not.

I believe you already figured this out, but just to be clear, the Collection I am using is initialized with capacity of 4 and beyond that, all new entries remove the oldest entry; it keeps a running list of the last 4 session handles.  The size shouldn't be unbounded because, as I understand it, there is no limit on the number of sessions that a single connection can have.

Reply:
I have never seen a code where we reuse a connection but open a new session. Is it possible in Hive? I would like to see an example of it before we build elaborate logging for it.

Reply:
Hey Peter, so it certainly is possible in Hive, the API exists, but I just looked at the JDBC code and the JDBC code does not allow new sessions on the same connection.  I'm a bit torn on it.  My gut tells me that that since the API allows it, we allow for it here.  What do you think?

Reply:
Here is the relevant code... it does not look like there are any error conditions barring a user from calling `openSession` multiple times on the same connection.  It just creates a new session and binds it to the connection.

https://github.com/apache/hive/blob/5ab73120605076c1afc65951bcca06045f532683/service/src/java/org/apache/hive/service/cli/thrift/ThriftCLIService.java#L337-L341

Reply:
I feel that the Hive code is complicated enough as it is. I would try to add complexity when is absolutely needed.
If we know of a customer workload where it is used like this, then go ahead and add the List. OTOH if we only find a non-trivial use-case for it, and it is not for correctness, but more for "possible help later" then I would try to avoid it.

Your thoughts?

Thanks,
Peter

Reply:
I think it's better to keep the current implementation; keep with the API.  I agree that Hive code is confusing, but I have included detailed comment that explains why I made this choice.  Does the comment make sense?

Reply:
@pvary I'm happy to do whatever is required to get an "Approved" so just let me know what will earn me that status :)

Reply:
TBH: I like consensus more than someone saying what is needed to get an "Approved" 😢 

That said I still think we do not need the `hiveSessionHandles` and the surrounding code.
Otherwise LGTM.

Thanks,
Peter